pkg/log/entry.go

```diff
@@ -1,66 +1,243 @@
 )
 
+// LogOps handles basic logging operations.
+type LogOps interface {
+	// Debug logs a debug message.
+	Debug(args ...interface{})
+
+	// Info logs an info message.
+	Info(args ...interface{})
+
+	// Warn logs a warning message.
+	Warn(args ...interface{})
+
+	// Error logs an error message.
+	Error(args ...interface{})
+
+	// Fatal logs a fatal message and terminates the program.
+	Fatal(args ...interface{})
+
+	// Panic logs a panic message and panics.
+	Panic(args ...interface{})
+
+	// Print logs a message at default level.
+	Print(args ...interface{})
+}
+
+// FormattedLogOps is an extension of LogOps that supports formatted logging.
+type FormattedLogOps interface {
+	// Debugf logs a debug message with formatting.
+	Debugf(format string, args ...interface{})
+
+	// Infof logs an info message with formatting.
+	Infof(format string, args ...interface{})
+
+	// Warnf logs a warning message with formatting.
+	Warnf(format string, args ...interface{})
+
+	// Errorf logs an error message with formatting.
+	Errorf(format string, args ...interface{})
+
+	// Fatalf logs a fatal message with formatting and terminates the program.
+	Fatalf(format string, args ...interface{})
+
+	// Panicf logs a panic message with formatting and panics.
+	Panicf(format string, args ...interface{})
+
+	// Printf logs a message with formatting at default level.
+	Printf(format string, args ...interface{})
+}
+
+// ContextualLogOps is a contextual logger that can be used to log messages with additional fields.
+type ContextualLogOps interface {
+	// WithFields returns a new Entry with the provided fields added.
+	WithFields(fields map[string]any) Entry
+
+	// WithField returns a new Entry with a single field added.
+	WithField(key string, value any) Entry
+
+	// WithError returns a new Entry with the error added to the fields.
+	WithError(err error) Entry
+
+	// WithContext returns a new Entry with the context added to the fields.
+	WithContext(ctx context.Context) Entry
+}
+
+// SystemLogger is an interface to handle system errors.
+type SystemLogger interface {
+	// SystemErr handles system errors with appropriate logging levels.
+	SystemErr(err error)
+
+	// WriterLevel returns an io.PipeWriter for the specified logging level.
+	WriterLevel(level slog.Level) *io.PipeWriter
+}
+
 // Entry is an abstraction to the
-// Logger and the logrus.Entry
+// Logger and the slog.Entry
 // so that *Logger always creates
 // an Entry copy which ensures no
 // Fields are being overwritten.
 type Entry interface {
-	// Basic Logging Operation
-	Debugf(format string, args ...any)
-	Infof(format string, args ...any)
-	Warnf(format string, args ...any)
-	Errorf(format string, args ...any)
-
-	// Attach contextual information to the logging entry
-	WithFields(fields map[string]any) Entry
-
-	// SystemErr is a method that disects the error
-	// and logs the appropriate level and fields for it.
-	SystemErr(err error)
+	LogOps
+	FormattedLogOps
+	ContextualLogOps
+	SystemLogger
 }
```

Comment:
We should keep this interface exactly the same and just implement it in terms of slog, why create all of the new methods above if they were not being used?

Reply:
Even if I  remove the rest, I will have to keep them in a separate interface and take those interfaces in entry since if I write all functions inside entry, the linter in the pipeline gives error for interfacebloat

Reply:
As for the extra methods, I can remove those but the problem for logrus vs slog levels and problems in Severity due to this still remains

Reply:
> As for the extra methods, I can remove those but the problem for logrus vs slog levels and problems in Severity due to this still remains

I'm not sure how they still remain, all you need to do is create a separate file that implements Entry in terms of log/slog?

Reply:
Here's an example: 

```
type slogger struct {
	l *slog.Logger
}

// Debugf implements Entry.
func (s *slogger) Debugf(format string, args ...any) {
	s.l.Debug(fmt.Sprintf(format, args...))
}

// Errorf implements Entry.
func (s *slogger) Errorf(format string, args ...any) {
	s.l.Error(fmt.Sprintf(format, args...))
}
```

For severity in the `errors` package, you can continue to use logrus levels, _or_ you can change it to slog, or better yet create an abstraction in the athens log package because the error package shouldn't import a third party lib

Reply:
An abstraction is preferred, imo.

Reply:
ok, got the abstraction part but still a bit confused, so you mean to say that i should only have the function declarations in entry and then implement those in log.go like it was done before my changes for SystemErr implementation or like for WithFields implementation?

Reply:
??

Reply:
I mean that you keep the interface exactly the same and just create a log/slog v1 implementation of that interface. Therefore, nothing changes outside of the log package (and the errors/severity, due to the abstraction we talked about above)